### PR TITLE
index: preserve extension parsing errors

### PR DESCRIPTION
**Pull request by tiennou**
Previously, we would clobber any extension-specific error message with an "extension is truncated" message. This makes `read_extension` correctly preserve those errors, takes responsibility for truncation errors, and adds a new message with the actual extension signature for unsupported mandatory extensions.

Inspired by #4857.

### CODE REVIEW

**Comment by ethomson** [comment]
The index changes look good to me - (the tests are failing on Linux though).

**Comment by tiennou** [comment]
I'm not sure why though. It seems the "splitindex" test repo I added isn't there ? If that makes any sense ?

Wait, the macOS build should be red as well, the test is failing there (at 16957).

**Comment by ethomson** [comment]
> Wait, the macOS build should be red as well, the test is failing there (at 16957).

Ugh, it certainly is.  I think the hoops that I had to jump through to get leaks to play nicely (it wants to background itself but keep some file descriptors open 😡) must not have worked right.  Sigh.

It's already very brittle (if you can't convince it to close that last file descriptor then it will just never exit and hang the builds), so I think I may have to disable leaks until I can make it more stable.

**Comment by tiennou** [comment]
I'm really confused by CI here, I can run that new testcase using the new test resource locally and I get a successful run, but CI fails because the resource is missing. Also, I suspect the Windows also has a problem with considering the build green while it failed…

**Comment by tiennou** [comment]
Rebased, was hoping to get CI to play nice but it still doesn't want to play along. I feel like I'm missing something, as I can't see anything wrong with the testcase setup…

**Comment by ethomson** [comment]
> Rebased, was hoping to get CI to play nice but it still doesn't want to play along. I feel like I'm missing something, as I can't see anything wrong with the testcase setup…

**Comment by tiennou** [comment]
OMG I've been wrangling with the repository open code and it didn't register that there were missing directories… Thanks for the pointer @ethomson, I've now rebased it with some `.gitkeep` files, and will take a note that a nicer validity message would be ❤️.

**Comment by ethomson** [comment]
> I don't think I have the permissions to do that (and I'm kinda evading doing that, because I'm a rebase-junkie and I like to keep it separated). I also was under the impression that it's only dependent on whether "Allow edit from maintainers" being checked ?

My workflow in `ethomson/foo` branches is to rebase also, so I wouldn't feel bad about that.  I'm not in front of my normal computer right now but we should get you these permissions.  Whether you choose to use them will remain up to you.  (It would make my life the tiniest bit easier, since I wouldn't have to set up your fork as a remote to fetch your PRs and build them locally, but that's not yet so common that I'm going to encourage you strongly to change your behavior.)

File: src/index.c [hunk 1]
```diff
@@ -2645,11 +2650,8 @@ static int parse_index(git_index *index, const char *buffer, size_t buffer_size)
 	while (buffer_size > INDEX_FOOTER_SIZE) {
 		size_t extension_size;
 
-		extension_size = read_extension(index, buffer, buffer_size);
-
-		/* see if we have read any bytes from the extension */
-		if (extension_size == 0) {
-			error = index_error_invalid("extension is truncated");
+		error = read_extension(&extension_size, index, buffer, buffer_size);
+		if (error < 0) {
```
Comment (pks-t) [inline src/index.c]: Minor nit: I'd reformat this to our usual style of

```
if ((error = read_extension(&extension_size, index, buffer, buffer_size)) < 0)
    goto done;
```

Please only consider this if you are already rerolling :)

**Comment by ethomson** [comment]
I'm 👍 on this.  @tiennou let me know if you want to make that change that @pks-t pointed out.  If not, we can go ahead and merge this.  If so, I'll hold off.

**Comment by tiennou** [comment]
Rebased, with style-fix applied. Thanks for the review!

**Pull request merged**